controllers/dopplersecret_controller_auth.go

```diff
@@ -0,0 +1,229 @@
+}
+
+// Handle OIDC authentication
+type OIDCAuthProvider struct {
+	oidcProvider *auth.OIDCAuthProvider
+	cacheKey     cache.Key
+	Log          logr.Logger
+}
+
+func (o *OIDCAuthProvider) GetAPIContext(ctx context.Context) (*api.APIContext, error) {
+	token, err := o.oidcProvider.GetToken(ctx)
+	if err != nil {
+		// On error, remove from cache to force retry
+		oidcProviderCache.Remove(o.cacheKey)
+		o.Log.Info("removed OIDC provider from cache due to error",
+			"namespace", o.oidcProvider.Namespace,
+			"error", err)
+		return nil, fmt.Errorf("unable to get OIDC token: %w", err)
+	}
+
+	return &api.APIContext{
+		Host:      o.oidcProvider.Host,
+		APIKey:    token,
+		VerifyTLS: o.oidcProvider.VerifyTLS,
+	}, nil
+}
+
+// Determine which authentication provider to use
+func (r *DopplerSecretReconciler) getAuthProvider(ctx context.Context, dopplerSecret *secretsv1alpha1.DopplerSecret) (AuthProvider, error) {
+	// Check what the token secret contains to determine auth type
+	tokenSecret := corev1.Secret{}
+	tokenNamespace := dopplerSecret.Namespace
+	if dopplerSecret.Spec.TokenSecretRef.Namespace != "" {
+		tokenNamespace = dopplerSecret.Spec.TokenSecretRef.Namespace
+	}
+
+	err := r.Client.Get(ctx, types.NamespacedName{
+		Name:      dopplerSecret.Spec.TokenSecretRef.Name,
+		Namespace: tokenNamespace,
+	}, &tokenSecret)
+	if err != nil {
+		return nil, fmt.Errorf("unable to fetch token secret: %w", err)
+	}
+
+	// Check what authentication fields exist
+	_, hasServiceToken := tokenSecret.Data["serviceToken"]
+	identity, hasIdentity := tokenSecret.Data["identity"]
+
+	// Ensure mutual exclusivity
+	if hasServiceToken && hasIdentity {
+		return nil, fmt.Errorf("token secret cannot contain both 'serviceToken' and 'identity' fields - use one or the other")
+	}
+
+	// Use OIDC authentication
+	if hasIdentity {
+		return r.createOIDCAuthProvider(dopplerSecret, string(identity), tokenSecret)
+	}
+
+	// Use service token authentication
+	if hasServiceToken {
+		return &ServiceTokenAuthProvider{
+			client:    r.Client,
+			tokenRef:  dopplerSecret.Spec.TokenSecretRef,
+			namespace: dopplerSecret.Namespace,
+			host:      dopplerSecret.Spec.Host,
+			verifyTLS: dopplerSecret.Spec.VerifyTLS,
+		}, nil
+	}
+
+	return nil, fmt.Errorf("token secret must contain either 'serviceToken' or 'identity' field")
+}
+
+// Create an OIDC authentication provider
+func (r *DopplerSecretReconciler) createOIDCAuthProvider(dopplerSecret *secretsv1alpha1.DopplerSecret, identity string, tokenSecret corev1.Secret) (AuthProvider, error) {
+
+	operatorNamespace, err := GetOwnNamespace()
+	if err != nil {
+		return nil, fmt.Errorf("unable to get operator namespace: %w", err)
+	}
+
+	// The dopplerTokenSecret audience allows us to validate that the token was issued
+	// for this specific identity
+	tokenSecretNamespace := dopplerSecret.Namespace
+	if dopplerSecret.Spec.TokenSecretRef.Namespace != "" {
+		tokenSecretNamespace = dopplerSecret.Spec.TokenSecretRef.Namespace
+	}
```

Comment:
nit/check: I'm sure it's fine if we pull it from the spec, but if we have the `tokenSecret` instance in scope already, could we just use that and not require this `if`?

Reply:
Also, it does look like an empty string is actually a valid value (equivalent to "default").